### PR TITLE
[ONNX] Add full checker mode in torch.onnx.export

### LINKED ISSUE #82589: [ONNX] Exporter successfully export a ONNX model, but unexecutable on ONNXRUNTIME due to failed shape inference..
### 🐛 Describe the bug

In exporter, there is a onnx.shape_inference_infer_shapes checker before exporting, but it is not using strict_mode as onnxruntime does before executing the model, so the discrepancy makes some of the models exportable, but not executable. 

A needed investigation of 
1. Is strict_mode necessary for onnxruntime?
2. Difference between passing strict_mode or not
3. What kind of model hitting this discrepancy

https://github.com/pytorch/pytorch/blob/ba84e9662e1c284805cce1652235eac1e4c56768/torch/csrc/jit/serialization/export.cpp#L1363-L1374

The following is a repro.

```python
import torch
from torch import Tensor
import io
from typing import Optional
import onnx
import onnxruntime as ort

class LoopNoneInput(torch.nn.Module):
    def forward(self, x):
        y: Optional[Tensor] = None
        for _ in range(x.size(0)):
            y = x
        return y


f = io.BytesIO()
x = torch.ones(1)
dynamic_axis_name = "condition"
torch.onnx.export(
    torch.jit.script(LoopNoneInput()),
    (x,),
    f,
    opset_version=16,
    # Ensure condition is not constant
    dynamic_axes={"x": {0: dynamic_axis_name}},
    input_names=["x"],
)

model = onnx.load_model_from_string(f.getvalue())
onnx.checker.check_model(model)

ort_input = {
    "x":x.cpu().numpy()
}
sess = ort.InferenceSession(model.SerializeToString() , ort.SessionOptions(), providers=['CPUExecutionProvider'])
out = sess.run(None, ort_inputs)
# Alternatively, one can use the following line to replace running ORT
# same outcome can be obtained.
#onnx.shape_inference.infer_shapes(model, strict_mode=True)
```

### Versions

nightly pytorch
nightly onnxruntime

### Relate Issues
[Input/Output type mismatch](https://github.com/onnx/onnx/issues/4424)

### PR DESCRIPTION
**Pull request by titaiwangms**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84789
* __->__ #83186

Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel @ezyang @gchanan

### CODE REVIEW

File: torch/csrc/jit/serialization/export.cpp [hunk 1]
```diff
@@ -1366,11 +1366,8 @@ void check_onnx_proto(const std::string& proto_string, bool full_check) {
     throw std::runtime_error("Invalid ONNX proto string.");
     return;
   }
-  onnx::checker::check_model(model);
-
-  if (full_check) {
-    onnx::shape_inference::InferShapes(model);
```
Comment (thiagocrepaldi) [inline torch/csrc/jit/serialization/export.cpp:1372]: why was this removed?

Comment (AllenTiTaiWang) [reply torch/csrc/jit/serialization/export.cpp:1372]: onnx::shape_inference::InferShapes(model) is included in onnx::checker::check_model(model) in the latest version.
https://github.com/onnx/onnx/blob/f0d824c0ead2825f7c24790c05f985ad1fb909f2/onnx/checker.cc#L941-L958

Comment (AllenTiTaiWang) [reply torch/csrc/jit/serialization/export.cpp:1372]: I thought check_model includes regular InferShapes, but it's not. I will later change it to:
```python
onnx::checker::check_model(model, full_check);
if (!full_check) {
  onnx::shape_inference::InferShapes(model);
}
```

File: torch/onnx/utils.py [hunk 2]
```diff
@@ -470,7 +471,10 @@ def forward(self, x):
             * ``True``: export all ``nn.Module`` forward calls as local function nodes.
             * Set of type of nn.Module: export ``nn.Module`` forward calls as local function nodes,
               only if the type of the ``nn.Module`` is found in the set.
-
+        full_model_check (bool, default False): if True, ONNX checker will give the exported model a
+            strict shape type inference check to make sure it's a valid ONNX model. Notice that only a
+            valid ONNX model can be executed on ONNXRUNTIME. Default to False, as users might run the
```
Comment (BowenBao) [inline torch/onnx/utils.py:476]: I wonder how it plays with `operator_export_type` that are not `ONNX`?

File: torch/onnx/utils.py [hunk 3]
```diff
@@ -200,6 +200,7 @@ def export(
     keep_initializers_as_inputs: Optional[bool] = None,
     custom_opsets: Optional[Mapping[str, int]] = None,
     export_modules_as_functions: Union[bool, Collection[Type[torch.nn.Module]]] = False,
+    full_model_check: bool = False,
```
Comment (BowenBao) [inline torch/onnx/utils.py:203]: We'd try avoid adding new arguments in an effort to not make the export api more complicated than it already is. Also a couple of alternatives to consider:

* Always run checker but have it emit warnings instead of errors. Removing the need of this argument.
* Always run checker but have it emit warnings instead of errors based on a new flag. With similar effect like "strict mode", this flag allows suppression of certain errors.

File: torch/onnx/utils.py (hunk 2 above)
Comment (AllenTiTaiWang) [reply torch/onnx/utils.py:476]: It won't check non-ONNX according to:
https://github.com/pytorch/pytorch/blob/6865a35c08d9e1ad7fd4daf481fa2803a8655f00/torch/onnx/utils.py#L1543-L1549

File: torch/onnx/utils.py (hunk 3 above)
Comment (AllenTiTaiWang) [reply torch/onnx/utils.py]: Done

**Comment by AllenTiTaiWang** [comment]
cc @BowenBao @shubhambhokare1

File: torch/csrc/jit/serialization/export.cpp (hunk 1 above)
Comment (AllenTiTaiWang) [reply torch/csrc/jit/serialization/export.cpp:1372]: Done

File: torch/csrc/jit/serialization/export.cpp [hunk 4]
```diff
@@ -1366,10 +1366,22 @@ void check_onnx_proto(const std::string& proto_string, bool full_check) {
...
+  // include non-strict model onnx::shape_inference::InferShapes(model).
+  // That's why we need to write another onnx::checker::check_model(model,
+  // full_check) here.
   if (full_check) {
-    onnx::shape_inference::InferShapes(model);
+    try {
+      onnx::checker::check_model(model, full_check);
+    } catch (const onnx::InferenceError& ex) {
+      TORCH_WARN(
+          "This is an invalid ONNX model with failed ONNX shape type inference. It is not executable by ONNXRUNTIME platform. More detail: ",
```
Comment (justinchuby) [inline torch/csrc/jit/serialization/export.cpp:1382]:
```suggestion
          "The exported ONNX model failed ONNX shape inference. The model will not be executable by the ONNX Runtime. If this is unintended and you believe there is a bug, please report an issue at https://github.com/pytorch/pytorch/issues. Error reported by the ONNX checker: ",
```

suggested wording (be sure to break the lines if too long)

File: torch/csrc/jit/serialization/export.cpp [hunk 5]
```diff
@@ -1366,10 +1366,22 @@ void check_onnx_proto(const std::string& proto_string, bool full_check) {
...
+  // full check including strict shape inference check
+  // Notice that onnx::checker::check_model doesn't
+  // include non-strict model onnx::shape_inference::InferShapes(model).
+  // That's why we need to write another onnx::checker::check_model(model,
+  // full_check) here.
   if (full_check) {
-    onnx::shape_inference::InferShapes(model);
+    try {
+      onnx::checker::check_model(model, full_check);
+    } catch (const onnx::InferenceError& ex) {
```
Comment (justinchuby) [inline torch/csrc/jit/serialization/export.cpp:1380]: Nice warning w/ original message

File: torch/csrc/onnx/init.cpp [hunk 6]
```diff
@@ -229,9 +229,10 @@ void initONNXBindings(PyObject* module) {
 
   m.def(
       "_check_onnx_proto",
-      [](const std::string& proto_string, bool full_check) {
-        check_onnx_proto(proto_string, full_check);
-      },
+      ::torch::wrap_pybind_function(
+          [](const std::string& proto_string, bool full_check) {
+            check_onnx_proto(proto_string, full_check);
+          }),
```
Comment (justinchuby) [inline torch/csrc/onnx/init.cpp:235]: Do we need a lambda here? Can this be `::torch::wrap_pybind_function(check_onnx_proto)`?

Just wondering, since there may be pybind limitations I am not aware of

File: torch/onnx/utils.py [hunk 7]
```diff
@@ -470,7 +470,6 @@ def forward(self, x):
             * ``True``: export all ``nn.Module`` forward calls as local function nodes.
             * Set of type of nn.Module: export ``nn.Module`` forward calls as local function nodes,
               only if the type of the ``nn.Module`` is found in the set.
-
```
Comment (justinchuby) [inline torch/onnx/utils.py:473]: I would put this line break back

File: torch/csrc/jit/serialization/export.cpp [hunk 8]
```diff
@@ -1366,10 +1366,22 @@ void check_onnx_proto(const std::string& proto_string, bool full_check) {
     throw std::runtime_error("Invalid ONNX proto string.");
     return;
   }
+  // baseline check
   onnx::checker::check_model(model);
-
+  onnx::shape_inference::InferShapes(model);
+  // full check including strict shape inference check
+  // Notice that onnx::checker::check_model doesn't
```
Comment (justinchuby) [inline torch/csrc/jit/serialization/export.cpp:1373]:
```suggestion
  // NOTE: onnx::checker::check_model doesn't
```

**Review by justinchuby — APPROVED** [review]

File: torch/csrc/jit/serialization/export.cpp [hunk 9]
```diff
@@ -1366,10 +1366,22 @@ void check_onnx_proto(const std::string& proto_string, bool full_check) {
...
+  // include non-strict model onnx::shape_inference::InferShapes(model).
+  // That's why we need to write another onnx::checker::check_model(model,
+  // full_check) here.
   if (full_check) {
```
Comment (justinchuby) [inline torch/csrc/jit/serialization/export.cpp:1377]: If full_check is always true, should we remove it from the function?

File: torch/csrc/jit/serialization/export.cpp (hunk 8 above)
Comment (justinchuby) [reply torch/csrc/jit/serialization/export.cpp:1373]: I would reword this to make the intention clear

**Review by justinchuby — CHANGES_REQUESTED** [review]
changes pending

File: torch/csrc/jit/serialization/export.cpp (hunk 4 above)
Comment (AllenTiTaiWang) [reply torch/csrc/jit/serialization/export.cpp]: Done

File: torch/csrc/onnx/init.cpp (hunk 6 above)
Comment (AllenTiTaiWang) [reply torch/csrc/onnx/init.cpp]: Tested, and it made no difference, so I modified it as suggested.

File: torch/onnx/utils.py (hunk 7 above)
Comment (AllenTiTaiWang) [reply torch/onnx/utils.py]: Done

File: torch/csrc/jit/serialization/export.cpp (hunk 8 above)
Comment (AllenTiTaiWang) [reply torch/csrc/jit/serialization/export.cpp]: Done

File: torch/csrc/jit/serialization/export.cpp (hunk 9 above)
Comment (AllenTiTaiWang) [reply torch/csrc/jit/serialization/export.cpp:1381]: Leaving this for now to see what others think.

**Review by justinchuby — APPROVED** [review]

File: torch/csrc/onnx/init.cpp (hunk 6 above)
Comment (AllenTiTaiWang) [reply torch/csrc/onnx/init.cpp:241]: Sorry, this is needed, since it passes parameter names to pybind and let users consume them in python. I think we should keep it.

**Comment by justinchuby** [comment]
Optional means a field can take None as an argument. If a paean is defaulted to True, we should annotate as `bool`, since None as an input is usually not meaningful

File: torch/csrc/onnx/init.cpp (hunk 6 above)
Comment (justinchuby) [reply torch/csrc/onnx/init.cpp:241]: Can you share a link if there's a doc?

File: torch/csrc/jit/serialization/export.cpp [hunk 10]
```diff
@@ -1370,10 +1370,26 @@ void check_onnx_proto(const std::string& proto_string, bool full_check) {
...
   }
+  // baseline check
   onnx::checker::check_model(model);
-
+  onnx::shape_inference::InferShapes(model);
+  // full check including strict shape inference check
+  // Note: onnx::checker::check_model doesn't
+  // include non-strict model onnx::shape_inference::InferShapes(model).
+  // That's why we need to write another onnx::checker::check_model(model,
+  // full_check) here.
```
Comment (justinchuby) [inline torch/csrc/jit/serialization/export.cpp:1380]: Just checking - was this reworded?

Comment (AllenTiTaiWang) [reply torch/csrc/jit/serialization/export.cpp:1380]: I think it's Notice that -> Note:

**Comment by AllenTiTaiWang** [comment]
> Optional means a field can take None as an argument. If a boolean is defaulted to True, we usually annotate it as `bool`, since None as an input is usually not meaningful

Is this comment actually for another PR?

File: torch/csrc/jit/serialization/export.cpp (hunk 10 above)
Comment (AllenTiTaiWang) [reply torch/csrc/jit/serialization/export.cpp:1380]: reword: 
1. clarify non-strict mode
2. call check_model twice reason

File: torch/csrc/jit/serialization/export.cpp (hunk 9 above)
Comment (thiagocrepaldi) [reply torch/csrc/jit/serialization/export.cpp:1381]: Why not use `torch.onnx._globals.py::_InternalGlobals.onnx_shape_inference` or similar to control this behavior?

Shape inference is not required by ONNX spec, it is an optional feature. If an invalid ONNX graph is being generated in ONNX converter, we should fix it, regardless of running or not inference shape every time

File: torch/csrc/onnx/init.cpp [hunk 11]
```diff
@@ -235,9 +235,10 @@ void initONNXBindings(PyObject* module) {
 
   m.def(
       "_check_onnx_proto",
-      [](const std::string& proto_string, bool full_check) {
-        check_onnx_proto(proto_string, full_check);
-      },
+      ::torch::wrap_pybind_function(
```
Comment (thiagocrepaldi) [inline torch/csrc/onnx/init.cpp:238]: Is this really needed? Many bindings in this file do not use. Try to understand what `::torch::wrap_pybind_function` does before start using it. If it is not needed, the lesser code, the better.

File: torch/csrc/onnx/init.cpp (hunk 6 above)
Comment (thiagocrepaldi) [reply torch/csrc/onnx/init.cpp:241]: > Do we need a lambda here? Can this be `::torch::wrap_pybind_function(check_onnx_proto)`?
> 
> Just wondering, since there may be pybind limitations I am not aware of

@justinchuby Do you know what `::torch::wrap_pybind_function(check_onnx_proto)`? Adding code we dont know what it does is dangerous. The safer bet is, if it is working and you dont know what the changes would cause, better not to touch it :)

**Review by thiagocrepaldi — CHANGES_REQUESTED** [review]
Just would like to make sure the offending operator is being fixed (in this PR or another) in addition to force a check that can break other backends

File: torch/csrc/onnx/init.cpp (hunk 11 above)
Comment (justinchuby) [reply torch/csrc/onnx/init.cpp:238]: Yes. See https://github.com/pytorch/pytorch/pull/78441

Comment (AllenTiTaiWang) [reply torch/csrc/onnx/init.cpp:238]: `::torch::wrap_pybind_function` helps to wrap the error into warning which help the PR achieve the goal of "not error out the exporter, but give the warning of why this model doesn't pass `strict_mode` shape type inference"

File: torch/csrc/jit/serialization/export.cpp [hunk 12]
```diff
@@ -1370,10 +1370,26 @@ void check_onnx_proto(const std::string& proto_string, bool full_check) {
     throw std::runtime_error("Invalid ONNX proto string.");
     return;
   }
+  // baseline check
   onnx::checker::check_model(model);
-
+  onnx::shape_inference::InferShapes(model);
```
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply torch/csrc/jit/serialization/export.cpp:1375]: > The thing is `stric_mode` is considered a baseline check now. If a model not passing `strict_mode` shape type inference check, it's not a valid ONNX graph.

That is my point. Why it is the baseline now? Why? It seems there is a bug in the conversion that is not being handled and we are waiting for shape inference to fix it for us.

We must export things correctly without dependencies on other APIs to fix a malformed operator. Maybe we need an actual example to see the error we are discussing

File: torch/csrc/jit/serialization/export.cpp [hunk 13]
```diff
@@ -1370,10 +1370,26 @@ void check_onnx_proto(const std::string& proto_string, bool full_check) {
...
+  // full check including strict shape inference check
+  // Note: onnx::checker::check_model doesn't
+  // include non-strict model onnx::shape_inference::InferShapes(model).
+  // That's why we need to write another onnx::checker::check_model(model,
+  // full_check) here.
   if (full_check) {
-    onnx::shape_inference::InferShapes(model);
+    try {
+      onnx::checker::check_model(model, full_check);
```
Comment (jcwchen) [inline torch/csrc/jit/serialization/export.cpp:1383]: I think we can directly run shape inference with strict mode:
```suggestion
      auto* schema_registry = onnx::OpSchemaRegistry::Instance();
      onnx::ShapeInferenceOptions options{
          /*check_type=*/True,
          /*error_mode=*/True};
      onnx::shape_inference::InferShapes(
          *model_proto, schema_registry, options);
```
Otherwise it runs onnx.checker twice.

File: torch/csrc/jit/serialization/export.cpp [hunk 14]
```diff
@@ -1370,10 +1370,26 @@ void check_onnx_proto(const std::string& proto_string, bool full_check) {
...
+    } catch (const onnx::InferenceError& ex) {
+      TORCH_WARN(
+          "The exported ONNX model failed ONNX shape inference."
+          "The model will not be executable by the ONNX Runtime."
+          "If this is unintended and you believe there is a bug,"
+          "please report an issue at https://github.com/pytorch/pytorch/issues."
+          "Error reported by the ONNX checker: ",
```
Comment (jcwchen) [inline torch/csrc/jit/serialization/export.cpp:1390]: More precise:
```suggestion
          "Error reported by the strict ONNX shape inference: ",
```

File: torch/csrc/jit/serialization/export.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (AllenTiTaiWang) [reply torch/csrc/jit/serialization/export.cpp:1375]: Resolved in offline meeting. This PR simply adds warning message, nit changing the original behavior of `check_onnx_proto`

**Review by jcwchen — APPROVED** [review]

File: torch/csrc/jit/serialization/export.cpp (hunk 13 above)
Comment (AllenTiTaiWang) [reply torch/csrc/jit/serialization/export.cpp]: Nice catch

File: torch/csrc/jit/serialization/export.cpp (hunk 14 above)
Comment (AllenTiTaiWang) [reply torch/csrc/jit/serialization/export.cpp]: Done

**Comment by AllenTiTaiWang** [comment]
> Just would like to make sure the offending operator is being fixed (in this PR or another) in addition to force a check that can break other backends

This warning message will be helpful for this case https://github.com/pytorch/pytorch/issues/83600 (the fix https://github.com/pytorch/pytorch/pull/83599)
With the warning message, we can identify the failed shape type inference ahead of onnxruntime.

**Review by thiagocrepaldi — APPROVED** [review]

**Review by BowenBao — APPROVED** [review]

**Comment by wschin** [comment]
Maybe let's use exporting mode to add a new mode to support the new evil mode?

**Comment by AllenTiTaiWang** [comment]
> Maybe let's use exporting mode to add a new mode to support the new evil mode?

To avoid the confusion, I will have another PR doing this. This further work can be track by #84884

**Comment by thiagocrepaldi** [comment]
Is there anything blocking this PR? AFAIK, the final version should 1) keep the current behavior and 2) print a useful warning for ONNX Runtime users, allowing them to identify the runtime issue.

Is that it?

**Comment by AllenTiTaiWang** [comment]
This PR needs ONNX version bump.

**Comment by titaiwangms** [comment]
> ## Merge failed
> **Reason**: GraphQL query fragment PRCheckSuites on CheckSuiteConnection { edges { node { app { name databaseId } workflowRun { workflow { name } url } checkRuns(first: 50) { nodes { name conclusion detailsUrl } pageInfo { endCursor hasNextPage } } conclusion } cursor } pageInfo { hasNextPage } }
> 
> query ($owner: String!, $name: String!, $number: Int!, $cursor: String!) { repository(name: $name, owner: $owner) { pullRequest(number: $number) { commits(last: 1) { nodes { commit { oid checkSuites(first: 10, after: $cursor) { ...PRCheckSuites } } } } } } } , args {'name': 'pytorch', 'owner': 'pytorch', 'number': 83186, 'cursor': 'Y3Vyc29yOnYyOpHPAAAAAoddcug='} failed: [{'message': 'Something went wrong while executing your query. Please include `07C3:7CCA:43256AB:88A6B31:63E4169C` when reporting this issue.'}]
> 
> Details for Dev Infra team

Hi @kit1980,

Do you know what issue this might be?

**Comment by kit1980** [comment]
Looks like GitHub hiccup?
Let's try again.

File: torch/csrc/jit/serialization/export.cpp [hunk 15]
```diff
@@ -1380,16 +1380,36 @@ std::string serialize_model_proto_to_string(
   return model_proto->SerializeAsString();
 }
 
-void check_onnx_proto(const std::string& proto_string, bool full_check) {
+void check_onnx_proto(const std::string& proto_string) {
```
Comment (kit1980) [inline torch/csrc/jit/serialization/export.cpp:1383]: This changes public API, no?

Comment (titaiwangms) [reply torch/csrc/jit/serialization/export.cpp:1383]: I think it does.  It's one of TORCH API. Is there anything else I should do?

Comment (kit1980) [reply torch/csrc/jit/serialization/export.cpp:1383]: You need to follow https://github.com/pytorch/pytorch/wiki/PyTorch's-Python-Frontend-Backward-and-Forward-Compatibility-Policy
It basically says first emit a warning for 2 releases, then can delete.
But for this function specifically I'm not sure if it's even used by anyone.
We may need to revert if there are complains.

Comment (titaiwangms) [reply torch/csrc/jit/serialization/export.cpp:1383]: Thanks!

Comment (kit1980) [reply torch/csrc/jit/serialization/export.cpp:1383]: I've added bc breaking labels.
